File: Leetcode/2002/2002.cpp
constexpr int N = 12;
constexpr int MAXN = 4096+5;

vector<pair<int,int>> isPalin;
bitset<N> bs;
int n;

bool f(string &s, int &length) {
    string str;
    for (int i = 0; i < n; ++i) {
        if (bs[i]) {
            str.push_back(s[i]);
        }
    }
    if (str.empty()) {
        return false;
    }
    length = (int)str.size();
    int halfLength = length >> 1;
    for (int i = 0; i < halfLength; ++i) {
        if (str[i] != str[length - 1 - i]) {
            return false;
        }
    }
    return true;
}

class Solution {
public:
    int maxProduct(string s) {
        vector<pair<int,int>>().swap(isPalin);
        bs.reset();
        n = (int)s.size();
        int total = 1<<n;
        int temp;
        for (int i = 0; i <= total; ++i) {
            if (f(s, temp)) {
                isPalin.emplace_back(i, temp);
            }
            bs = bitset<N>(bs.to_ulong()+1LL);
        }
        int ans = 0;
        int length = (int)isPalin.size();
        for (int i = 0; i < length-1; ++i) {
            for (int j = length - 1; j > i; --j) {
                if (!(isPalin[i].first & isPalin[j].first)) {
                    ans = max(ans, isPalin[i].second * isPalin[j].second);
                }
            }
        }
        return ans;
    }
};

File: Leetcode/2002/2002.go
type Node struct {
	pos, length int
}

const N = 13
var bitset [N]bool

func f(val int) {
	for i := 0; i < N; i++ {
		bitset[i] = false
	}
	var index = 0
	for val > 0 {
		if (val & 1) != 0 {
			bitset[index] = true
		}
		index++
		val /= 2
	}
}

func check(s *string, length *int, val int) bool {
	var str []rune
	f(val)
	for i := 0; i < len(*s); i++ {
		if bitset[i] {
			str = append(str, rune((*s)[i]))
		}
	}
	if len(str) == 0 {
		return false
	}
	halfLength := len(str) >> 1
	for i := 0; i < halfLength; i++ {
		if str[i] != str[len(str)-i-1] {
			return false
		}
	}
	*length = len(str)
	return true
}

func maxProduct(s string) int {
	var isPalin []Node
	total := 1<<len(s)
	var length int
	for i := 0; i <= total; i++ {
		if check(&s, &length, i) {
			isPalin = append(isPalin, Node{length: length,pos: i})
		}
	}
    ans := 0
	for i := 0; i < len(isPalin)-1; i++ {
		for j := i+1; j < len(isPalin); j++ {
			if (isPalin[i].pos & isPalin[j].pos) == 0 {
				temp := isPalin[i].length * isPalin[j].length
				if temp > ans {
					ans = temp
				}
			}
		}
	}
	return ans
}